{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Normalization toolbox"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Why normalize your data at all? So that your data is comparable. There are many sources of variation in data processing pipelines and (metabol)omics data is also variable. The differences in concentrations between individual metabolites or individual samples can both be very high. Good normalization methods need to preserve these ratios and make sure that neither low- nor highly abundant metabolites are disadvantaged in the process [(Kohl 2012)](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3337420/pdf/11306_2011_Article_350.pdf)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[This paper here](https://www.nature.com/articles/srep38881.pdf) is a great study that compares different normalization methods for LC/MS bases metabolomics. I'd use it as a starting point if we decide to do one for FIA-MS"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Note to self:\n",
    "here is more literature to include:\n",
    "- Impact of normalization on the outcome [(Saccenti 2017)](https://pubs.acs.org/doi/10.1021/acs.jproteome.6b00704)\n",
    "- Novel developments [(Engel 2013)](https://www.sciencedirect.com/science/article/abs/pii/S0165993613001465)\n",
    "- Normalization in metabolomics [(Misra 2020)](https://journals.sagepub.com/doi/abs/10.1177/1469066720918446?journalCode=emsa)\n",
    "- A Python-Based Pipeline for Preprocessing LC–MSData for Untargeted Metabolomics Workflows [(Riquelme 2020)](https://www.mdpi.com/2218-1989/10/10/416)\n",
    "- Hints about PLR (don't know how to implement it), [.](http://www.compositionaldata.com/codawork2017/abstracts/9thFriday/Oral/OMI-2-2.pdf) [.](https://www.sciencedirect.com/science/article/abs/pii/S0021967314012990?via%3Dihub) [.](https://www.sciencedirect.com/science/article/abs/pii/S0003267012005685)\n",
    "- Batch effect [(Gagnon-Bartsch and Speed 2012)](https://watermark.silverchair.com/kxr034.pdf?token=AQECAHi208BE49Ooan9kkhW_Ercy7Dm3ZL_9Cf3qfKAc485ysgAAArYwggKyBgkqhkiG9w0BBwagggKjMIICnwIBADCCApgGCSqGSIb3DQEHATAeBglghkgBZQMEAS4wEQQMipKBNk2L1l3QyfmiAgEQgIICaaHlZyuwWjWC5T5v2hDtYSKaPIiYhhbr0szzbcrp3xtOaziuxhzkB0NFyJJhE37sGWeBsuyKdyswTmtfJTlPcLDkYP-vvXv9ykjyulkCuTRQ2MHivovM1UkGWB5AnQvM-ZFcokabvt9JrG8j6xXzTJmE1fmifs8KCzIl0yC9i5V9Py7ctlS4odzj5uYKEqqSRceEyZk_QAFwtr9AUrqwP-q9MUZ3Wh28JLEHcKgVUFyNHbnk-r2Or2ouIdUDkwwQnRTPXZDfrwLAASs0E5yir1Kdsi4_Jq7m1klvinDAaYEuxiHo7O1AjOs5U-ltPbh2sQUrhWGvanWoaH_6fXFIviKsuRssL7FCaxzOsbbdY7UAB13pRAiHIHGu7-qhpX4hakVT_OGTF5dD2q8O6tVibfBBzoLFvIjbMYLFGZWhHXxLnfnTjFDuojdo6QuVefpWrjHEHHg3seCPFJx2Yv3Z_hu8CvRJU1VMwwCQ5SmCjBh7_TY8MpbKpj3gcR85wj0akjq_gd02RL8LMpvjVFu1rb1-KpZ__bokI1DN02pOGT-WJl2ot7l9xqUyF9u458MMHQ5xc0CvzNMVCrmppFtGzCTtB_z6AgkXzseaGZrTqlGt-JOWbKev7de9BRM58HCYIxV5zIThZoZkPEUE2-cZOV_Klc3FvEN2i1eoyYIASH8iODuwdsDm8RcjjECecgSPG31dfO9HMFUjaOUWGSK40VcTBuc9OeNzuHOybIEAGsgYQ5XEO8R3R9pMvMi7KcsxK0xQv6SSwX6AjVQJnkVlCdaQObxMMOYqMavxuJPNoYYAdDZs2tbifRag), [(Cheng Li and Rabinovic)](https://watermark.silverchair.com/kxj037.pdf?token=AQECAHi208BE49Ooan9kkhW_Ercy7Dm3ZL_9Cf3qfKAc485ysgAAArQwggKwBgkqhkiG9w0BBwagggKhMIICnQIBADCCApYGCSqGSIb3DQEHATAeBglghkgBZQMEAS4wEQQMtFy6c03qWXwVaMgwAgEQgIICZ7O9atjsPumn5RrucSZ4xM9V9Z9CYeM59bDNp7tD_7iZVAu9d9ODm2gl7PH--EMwgoMO6EJ6H0uQ3JUojod0dbiEbxfjv0ELlZCnPg1X9akLZB_yRJE7aPAE7MRr6I0hF8_374Z2irAuvr7mjDQ_bZ40a1tujXvwwxMztgn4ozKC6sL1NLj0Uhs2KYhJrjHed2be4gwsARFjjv0FRzwzIR81ov2dMq3-qtBdgmSWqMzsuKLE0qY4VWH_trPDg393UWGhw5ZdyyOaLts70hUNL22-7qDLSRdbTxJs_IkKxOVDfdrrPTNADm1TRPyAWlT7nsaOuLtNguH4fxj-gBx31PTuYnyiOgK1LXy5gSJtgoSslyL3dyNtUU9ttNjnALYX9-YXjWTCtgt8pByuJiupy9b6xc_IPfDdnK4Ef7PfcZLXYJUjtlLoxyJdhbAquYvr-KjL8l0wqg0BDyrrL5fML8GNh3-ss_d2ubg2QzgPgWAhFN1sOvfRJDXirtM_9MrK69s_LIo5rdezrWQfqDotAXKq2J3mhg95viR_FH3sf3SG4PK2PKy5rEhKIXDuIuXIYqF1YD8n3DMDyt1ASpmH4JaluO8-8QnL-XZtIkbimkmbNg53cTD8rujcp-e0J47wbsMpHWNs3pCN_2SivxQ7jzSpT_9HDhdnBxH-uaBVj3xZkixvcb8wUDhsl4oxUpFwwXaWwzs-3ToKkadGbM0lWjIYWSUg7B34FmMDcUVdy4UgDrH8FS2QuhtK_uGB8_-FcBaixeloZfsk0QSEgK6rrGIcrVXC5WxGcOes8QyfesPAqwYxydwslg), [(Wherens 2016)](https://link.springer.com/article/10.1007/s11306-016-1015-8)\n",
    "- Internal Standard normalization taking care of Cross Contribution [(Redestig 2009)](https://pubs.acs.org/doi/abs/10.1021/ac901143w?casa_token=SMFtt8olvVEAAAAA:btSEUUFPXdGg4Cx90tG1DXWGDOiIXP0vrsQ0pmLDccitoP_amy8j-oRBDBJXHBkZ84jqgWmtIhPRokM)\n",
    "- Workflow for metabolomics [(De Livera)](https://pubs.acs.org/doi/10.1021/ac302748b) (it's from the Terry Speed group and he seems to be good at his job)\n",
    "- Surrogate variable analysis (SVA), not sure how relevant that is though [(Leek 2007)](https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.0030161)\n",
    "\n",
    "### to include:\n",
    "- constant sum\n",
    "- [quantile normalization](https://watermark.silverchair.com/190185.pdf?token=AQECAHi208BE49Ooan9kkhW_Ercy7Dm3ZL_9Cf3qfKAc485ysgAAArswggK3BgkqhkiG9w0BBwagggKoMIICpAIBADCCAp0GCSqGSIb3DQEHATAeBglghkgBZQMEAS4wEQQMW8dHBAPZhc2xZmniAgEQgIICbp3S-UhHiEsDmL2jlngSyxCN-rWd2LquzmRg58iAmZqt4Wi5UGEMpLC8dhD1sMovXXZfs09Ff_ae62aww8h_iIYWLEyVi-W19oDdljxe-FprdiBFq_bwEuNo2znrbzZnyH6hegGXUVdUrp0CfYpG1-H1pKlj3cxuF-bZGnhzB2J7fZ66fLz02w1CQM9WpuHgNG1R_VebtwOPPuJ_RMMoxbXxh5589w5si7tEuDnGu8vUy4Hh3W1caCbKwutQomBANDASYMQpjGFLmtHZlzhTdIPozbnue8oduA4urZoKt5WFlMegaNRmUybyfZQJNJ48SZshcefMTa-q5EE5xAC_kJmITGubMMKcBjUOH0KdXBfGFjb8OFi6XkwuZo2lfYtk5DyW08GQAip18uDiVANM2xkfldmWnnDbD264QAUZbtcFhSpTcCF7trHOGYHhNV-e36EMt3L14t2U7q_UGJVDkozXrxf-eh_m4Cl2w0yB8H6OptK3zdSjH77iLBxMr2SI79ZX77H4WWkwyyH22qjXCVJXVUuHEFAo6Kbcttrjt2IFHMot9bcLbsuetfATjYVVU0OWtGyhYHefkmkYsE2iJ1mrt8EEwNn9PgfqCBiXHLhBMCM4ML6jNByT6QDTSPp9gXAThQO8LVf3pvGzaog1FBQ7br9-riLHkd4Ub8AlL3f1yIn42Ec632VpaQBc-S6rI_vn4qV-1hbf3Xy1_h9s2u4iH4mS-EJueu0VUO4QSsWi8wVknKA8sc1zcBrPDLH-ektOWvtl3kw7zi6xSJoM6JHyF3nIJAKpwCWYMVtja4qnXGCknYVgaM3tsB3r95k) (reviewed in Kohl, but only really good for large 50+ datasets) \n",
    "- Pairwise log-ratios\n",
    "\n",
    "### to include because identified to be great for LC/MS data in [this paper here](https://www.nature.com/articles/srep38881.pdf):\n",
    "- Log transform\n",
    "- VSN\n",
    "\n",
    "### to include because identified to be great for GC/MS data in [this paper here](https://bmcgenomics.biomedcentral.com/articles/10.1186/1471-2164-7-142):\n",
    "- Auto Scaling\n",
    "- Range Scaling"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Re-write this part if you want to use it becuase it's mainly copied\n",
    "\n",
    "[(Vu et. al 2017)](http://bionmr.unl.edu/files/publications/152.pdf) compared 9 different normalization methods for NMR metabolomics data by adding Gaussian noise and dilution factors and comparing their peak recovery. The methods tested were: probabilistic quotient (PQ) [(Dieterle et. al, 2006)](https://pubs.acs.org/doi/10.1021/ac051632c), histogram matching (HM) [(Torgrip et al. 2008)](https://link.springer.com/article/10.1007%2Fs11306-007-0102-2#auth-J_-Lindberg), standard normal variate (SNV) [(Barnes et al. 1989)](https://journals.sagepub.com/doi/abs/10.1366/0003702894202201?casa_token=Ywnp29pqbS0AAAAA:_peitRZdEDIBNjf28GvJaHHBlGDzJJlVQejdthxBxMxWyo5wFllaXHh_F7Cff4mfdXcb1IEgB7kLAQ), multiplicative scatter correction (MSC) [(Windig et al. 2008)](https://journals.sagepub.com/doi/10.1366/000370208786049097), quantile (Q) [(Bolstad 2003)](https://watermark.silverchair.com/190185.pdf?token=AQECAHi208BE49Ooan9kkhW_Ercy7Dm3ZL_9Cf3qfKAc485ysgAAArswggK3BgkqhkiG9w0BBwagggKoMIICpAIBADCCAp0GCSqGSIb3DQEHATAeBglghkgBZQMEAS4wEQQMW8dHBAPZhc2xZmniAgEQgIICbp3S-UhHiEsDmL2jlngSyxCN-rWd2LquzmRg58iAmZqt4Wi5UGEMpLC8dhD1sMovXXZfs09Ff_ae62aww8h_iIYWLEyVi-W19oDdljxe-FprdiBFq_bwEuNo2znrbzZnyH6hegGXUVdUrp0CfYpG1-H1pKlj3cxuF-bZGnhzB2J7fZ66fLz02w1CQM9WpuHgNG1R_VebtwOPPuJ_RMMoxbXxh5589w5si7tEuDnGu8vUy4Hh3W1caCbKwutQomBANDASYMQpjGFLmtHZlzhTdIPozbnue8oduA4urZoKt5WFlMegaNRmUybyfZQJNJ48SZshcefMTa-q5EE5xAC_kJmITGubMMKcBjUOH0KdXBfGFjb8OFi6XkwuZo2lfYtk5DyW08GQAip18uDiVANM2xkfldmWnnDbD264QAUZbtcFhSpTcCF7trHOGYHhNV-e36EMt3L14t2U7q_UGJVDkozXrxf-eh_m4Cl2w0yB8H6OptK3zdSjH77iLBxMr2SI79ZX77H4WWkwyyH22qjXCVJXVUuHEFAo6Kbcttrjt2IFHMot9bcLbsuetfATjYVVU0OWtGyhYHefkmkYsE2iJ1mrt8EEwNn9PgfqCBiXHLhBMCM4ML6jNByT6QDTSPp9gXAThQO8LVf3pvGzaog1FBQ7br9-riLHkd4Ub8AlL3f1yIn42Ec632VpaQBc-S6rI_vn4qV-1hbf3Xy1_h9s2u4iH4mS-EJueu0VUO4QSsWi8wVknKA8sc1zcBrPDLH-ektOWvtl3kw7zi6xSJoM6JHyF3nIJAKpwCWYMVtja4qnXGCknYVgaM3tsB3r95k), natural cubic splines (CSpline) [(Workman et al. 2002)](https://genomebiology.biomedcentral.com/articles/10.1186/gb-2002-3-9-research0048), smoothing splines (SSpline) [(Fujioka and Kano 2005)](https://magnus.ece.gatech.edu/Papers/BSplinesIJICIC05.pdf), constant sum (CS) and region of interest (ROI) [(Dieterle et. al, 2006)](https://pubs.acs.org/doi/10.1021/ac051632c). At low noise levels, all of the methods except for HM performed similarily well. The three best performing methods were PQ, CS and ROI, also for high dilution factors. At high variance levels on the other hand ROI dropped off. Thus they recommend using either PQ or CS whenever possible. The two methods displayed some differences as well; CS performing slightly better at high noise levels while PQ performed significantly better than CS at low to moderate noise levels."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is an example notebook for the normalization module. In general, the module provides different custom functions to normalize your data in different ways. Normalization is a tricky subject and the approach you choose can impact your results, please be aware of that. Some considerations are discussed here in the notebook."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Data used: AUT-440-446_FIAMSMatrixLinearityR2-Reproducibility_DataProcessing\\20201223_AddAdducts-SumInt-BackSub\n",
    "\n",
    "This data set is blank corrected"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import copy\n",
    "import pickle\n",
    "import BFAIR.normalization as normalization"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"data/normalization_Data/Ecoli_intensities_linearity.txt\", 'rb') as handle:\n",
    "    intensities_Ecoli = pickle.loads(handle.read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_group_name</th>\n",
       "      <th>Metabolite</th>\n",
       "      <th>Formula</th>\n",
       "      <th>Intensity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Linearity_P1Ecoli_100xDil_1</td>\n",
       "      <td>12dgr120_c</td>\n",
       "      <td>C27H52O5</td>\n",
       "      <td>9.583652e+03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Linearity_P1Ecoli_100xDil_1</td>\n",
       "      <td>12dgr140_c</td>\n",
       "      <td>C31H60O5</td>\n",
       "      <td>1.276414e+03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Linearity_P1Ecoli_100xDil_1</td>\n",
       "      <td>12dgr141_c</td>\n",
       "      <td>C31H56O5</td>\n",
       "      <td>1.282265e+03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Linearity_P1Ecoli_100xDil_1</td>\n",
       "      <td>12dgr160_c</td>\n",
       "      <td>C35H68O5</td>\n",
       "      <td>2.968548e+04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Linearity_P1Ecoli_100xDil_1</td>\n",
       "      <td>12dgr161_c</td>\n",
       "      <td>C35H64O5</td>\n",
       "      <td>2.144749e+04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10388</th>\n",
       "      <td>Linearity_P1Ecoli_1xDil_3</td>\n",
       "      <td>xdp_c</td>\n",
       "      <td>C10H14N4O12P2</td>\n",
       "      <td>1.575200e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10389</th>\n",
       "      <td>Linearity_P1Ecoli_1xDil_3</td>\n",
       "      <td>xmp_c</td>\n",
       "      <td>C10H13N4O9P1</td>\n",
       "      <td>9.985480e+04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10390</th>\n",
       "      <td>Linearity_P1Ecoli_1xDil_3</td>\n",
       "      <td>xtp_c</td>\n",
       "      <td>C10H15N4O15P3</td>\n",
       "      <td>3.546930e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10391</th>\n",
       "      <td>Linearity_P1Ecoli_1xDil_3</td>\n",
       "      <td>xtsn_c</td>\n",
       "      <td>C10H12N4O6</td>\n",
       "      <td>9.593014e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10392</th>\n",
       "      <td>Linearity_P1Ecoli_1xDil_3</td>\n",
       "      <td>zn2_c</td>\n",
       "      <td>Zn</td>\n",
       "      <td>5.471120e+04</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10393 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 sample_group_name  Metabolite        Formula     Intensity\n",
       "0      Linearity_P1Ecoli_100xDil_1  12dgr120_c       C27H52O5  9.583652e+03\n",
       "1      Linearity_P1Ecoli_100xDil_1  12dgr140_c       C31H60O5  1.276414e+03\n",
       "2      Linearity_P1Ecoli_100xDil_1  12dgr141_c       C31H56O5  1.282265e+03\n",
       "3      Linearity_P1Ecoli_100xDil_1  12dgr160_c       C35H68O5  2.968548e+04\n",
       "4      Linearity_P1Ecoli_100xDil_1  12dgr161_c       C35H64O5  2.144749e+04\n",
       "...                            ...         ...            ...           ...\n",
       "10388    Linearity_P1Ecoli_1xDil_3       xdp_c  C10H14N4O12P2  1.575200e+06\n",
       "10389    Linearity_P1Ecoli_1xDil_3       xmp_c   C10H13N4O9P1  9.985480e+04\n",
       "10390    Linearity_P1Ecoli_1xDil_3       xtp_c  C10H15N4O15P3  3.546930e+05\n",
       "10391    Linearity_P1Ecoli_1xDil_3      xtsn_c     C10H12N4O6  9.593014e+05\n",
       "10392    Linearity_P1Ecoli_1xDil_3       zn2_c             Zn  5.471120e+04\n",
       "\n",
       "[10393 rows x 4 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intensities_Ecoli"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Normalization methods"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### min/max feature scaling: value - min / max - min"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A standard method to scale your data between 0 (min) and 1 (max). This method depends soley on the smallest and the largest value in your dataset, so the results can easily be influenced by dominant elements (e.g. very highly abundant metabolites) or outliers. This method should probably only be used if your data has to be scaled from 0 to 1."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_group_name</th>\n",
       "      <th>Metabolite</th>\n",
       "      <th>Formula</th>\n",
       "      <th>Intensity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Linearity_P1Ecoli_100xDil_1</td>\n",
       "      <td>12dgr120_c</td>\n",
       "      <td>C27H52O5</td>\n",
       "      <td>1.416323e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Linearity_P1Ecoli_100xDil_1</td>\n",
       "      <td>12dgr140_c</td>\n",
       "      <td>C31H60O5</td>\n",
       "      <td>4.546221e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Linearity_P1Ecoli_100xDil_1</td>\n",
       "      <td>12dgr141_c</td>\n",
       "      <td>C31H56O5</td>\n",
       "      <td>4.642775e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Linearity_P1Ecoli_100xDil_1</td>\n",
       "      <td>12dgr160_c</td>\n",
       "      <td>C35H68O5</td>\n",
       "      <td>4.733527e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Linearity_P1Ecoli_100xDil_1</td>\n",
       "      <td>12dgr161_c</td>\n",
       "      <td>C35H64O5</td>\n",
       "      <td>3.374094e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10388</th>\n",
       "      <td>Linearity_P1Ecoli_1xDil_3</td>\n",
       "      <td>xdp_c</td>\n",
       "      <td>C10H14N4O12P2</td>\n",
       "      <td>5.051276e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10389</th>\n",
       "      <td>Linearity_P1Ecoli_1xDil_3</td>\n",
       "      <td>xmp_c</td>\n",
       "      <td>C10H13N4O9P1</td>\n",
       "      <td>3.167472e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10390</th>\n",
       "      <td>Linearity_P1Ecoli_1xDil_3</td>\n",
       "      <td>xtp_c</td>\n",
       "      <td>C10H15N4O15P3</td>\n",
       "      <td>1.134548e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10391</th>\n",
       "      <td>Linearity_P1Ecoli_1xDil_3</td>\n",
       "      <td>xtsn_c</td>\n",
       "      <td>C10H12N4O6</td>\n",
       "      <td>3.074797e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10392</th>\n",
       "      <td>Linearity_P1Ecoli_1xDil_3</td>\n",
       "      <td>zn2_c</td>\n",
       "      <td>Zn</td>\n",
       "      <td>1.718768e-05</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10393 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 sample_group_name  Metabolite        Formula     Intensity\n",
       "0      Linearity_P1Ecoli_100xDil_1  12dgr120_c       C27H52O5  1.416323e-05\n",
       "1      Linearity_P1Ecoli_100xDil_1  12dgr140_c       C31H60O5  4.546221e-07\n",
       "2      Linearity_P1Ecoli_100xDil_1  12dgr141_c       C31H56O5  4.642775e-07\n",
       "3      Linearity_P1Ecoli_100xDil_1  12dgr160_c       C35H68O5  4.733527e-05\n",
       "4      Linearity_P1Ecoli_100xDil_1  12dgr161_c       C35H64O5  3.374094e-05\n",
       "...                            ...         ...            ...           ...\n",
       "10388    Linearity_P1Ecoli_1xDil_3       xdp_c  C10H14N4O12P2  5.051276e-04\n",
       "10389    Linearity_P1Ecoli_1xDil_3       xmp_c   C10H13N4O9P1  3.167472e-05\n",
       "10390    Linearity_P1Ecoli_1xDil_3       xtp_c  C10H15N4O15P3  1.134548e-04\n",
       "10391    Linearity_P1Ecoli_1xDil_3      xtsn_c     C10H12N4O6  3.074797e-04\n",
       "10392    Linearity_P1Ecoli_1xDil_3       zn2_c             Zn  1.718768e-05\n",
       "\n",
       "[10393 rows x 4 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "normalization.min_max_norm(intensities_Ecoli, columnname='Intensity')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Total Sum Intensity, value / tsi"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The following methods are all based on total sum normalization, which comes with caveats. One should keep in mind that this depends on the intensity of all the measured metabolites. If there is an unequal distribution of values or a particularly dominant metabolite (as can easily be the case in metabolic engineering) then this would alter the normalized concentration of the other metabolites (a known problem with closed data, [(Filzmoser & Walczak, 2014)](https://www.sciencedirect.com/science/article/abs/pii/S0021967314012990?via%3Dihub)). Other methods such as PQN, which makes use of TSI in its initial step are recommended [(Dieterle et. al, 2006)](https://pubs.acs.org/doi/10.1021/ac051632c). Also, while this method preserves the overall signal quality, the can skew correlations between signals [(Craig et. al, 2006)](https://pubs.acs.org/doi/10.1021/ac0519312), and intensities for the same compound in different KOs can appear different."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_group_name</th>\n",
       "      <th>Metabolite</th>\n",
       "      <th>Formula</th>\n",
       "      <th>Intensity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Linearity_P1Ecoli_100xDil_1</td>\n",
       "      <td>12dgr120_c</td>\n",
       "      <td>C27H52O5</td>\n",
       "      <td>1.401056e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Linearity_P1Ecoli_100xDil_1</td>\n",
       "      <td>12dgr140_c</td>\n",
       "      <td>C31H60O5</td>\n",
       "      <td>1.866019e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Linearity_P1Ecoli_100xDil_1</td>\n",
       "      <td>12dgr141_c</td>\n",
       "      <td>C31H56O5</td>\n",
       "      <td>1.874573e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Linearity_P1Ecoli_100xDil_1</td>\n",
       "      <td>12dgr160_c</td>\n",
       "      <td>C35H68O5</td>\n",
       "      <td>4.339788e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Linearity_P1Ecoli_100xDil_1</td>\n",
       "      <td>12dgr161_c</td>\n",
       "      <td>C35H64O5</td>\n",
       "      <td>3.135457e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10388</th>\n",
       "      <td>Linearity_P1Ecoli_1xDil_3</td>\n",
       "      <td>xdp_c</td>\n",
       "      <td>C10H14N4O12P2</td>\n",
       "      <td>5.544914e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10389</th>\n",
       "      <td>Linearity_P1Ecoli_1xDil_3</td>\n",
       "      <td>xmp_c</td>\n",
       "      <td>C10H13N4O9P1</td>\n",
       "      <td>3.515023e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10390</th>\n",
       "      <td>Linearity_P1Ecoli_1xDil_3</td>\n",
       "      <td>xtp_c</td>\n",
       "      <td>C10H15N4O15P3</td>\n",
       "      <td>1.248567e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10391</th>\n",
       "      <td>Linearity_P1Ecoli_1xDil_3</td>\n",
       "      <td>xtsn_c</td>\n",
       "      <td>C10H12N4O6</td>\n",
       "      <td>3.376870e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10392</th>\n",
       "      <td>Linearity_P1Ecoli_1xDil_3</td>\n",
       "      <td>zn2_c</td>\n",
       "      <td>Zn</td>\n",
       "      <td>1.925908e-06</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10393 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 sample_group_name  Metabolite        Formula     Intensity\n",
       "0      Linearity_P1Ecoli_100xDil_1  12dgr120_c       C27H52O5  1.401056e-06\n",
       "1      Linearity_P1Ecoli_100xDil_1  12dgr140_c       C31H60O5  1.866019e-07\n",
       "2      Linearity_P1Ecoli_100xDil_1  12dgr141_c       C31H56O5  1.874573e-07\n",
       "3      Linearity_P1Ecoli_100xDil_1  12dgr160_c       C35H68O5  4.339788e-06\n",
       "4      Linearity_P1Ecoli_100xDil_1  12dgr161_c       C35H64O5  3.135457e-06\n",
       "...                            ...         ...            ...           ...\n",
       "10388    Linearity_P1Ecoli_1xDil_3       xdp_c  C10H14N4O12P2  5.544914e-05\n",
       "10389    Linearity_P1Ecoli_1xDil_3       xmp_c   C10H13N4O9P1  3.515023e-06\n",
       "10390    Linearity_P1Ecoli_1xDil_3       xtp_c  C10H15N4O15P3  1.248567e-05\n",
       "10391    Linearity_P1Ecoli_1xDil_3      xtsn_c     C10H12N4O6  3.376870e-05\n",
       "10392    Linearity_P1Ecoli_1xDil_3       zn2_c             Zn  1.925908e-06\n",
       "\n",
       "[10393 rows x 4 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "normalization.tsi_norm(intensities_Ecoli, columnname='Intensity')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## These methods are based on the E. coli model that we used in the INCA example for now; [iJS2012](https://bmcsystbiol.biomedcentral.com/articles/10.1186/1752-0509-6-9).\n",
    "Biomass reactions and the corresponding metabolite identifiers will be added for different organisms. All metabolites that are part of the biomass function, substrates and products, are being considered"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Main Biomass components"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "biomass_mets = ['phe__L_c', 'mlthf_c' ,'oaa_c', 'lys__L_c',\n",
    "                'atp_c', 'ser__L_c', 'g3p_c', 'tyr__L_c', 'pep_c',\n",
    "                'met__L_c', 'g6p_c', 'akg_c', 'glu__L_c',\n",
    "                'gln__L_c', 'r5p_c', 'f6p_c', 'pyr_c', 'gly_c',\n",
    "                'thr_c', 'asp__L_c', 'nadph_c', 'cys__L_c',\n",
    "                '3pg_c', 'val__L_c', 'ala__L_c', 'ile__L_c',\n",
    "                'asn__L_c', 'his__L_c', 'leu__L_c', 'accoa_c',\n",
    "                'arg__L_c', 'pro__L_c', 'trp__L_c', 'nadh_c']\n",
    "bm_vals = [0.176, 0.443, 0.34, 0.326, 33.247, 0.205, 0.129, \n",
    "           0.131, 0.051, 0.146, 0.205, 0.087, 0.25, 0.25, 0.754,\n",
    "           0.071, 0.083, 0.582, 0.241, 0.229, 5.363, 0.087, 0.619,\n",
    "           0.402, 0.488, 0.276, 0.229, 0.09, 0.428, 2.51, 0.281,\n",
    "           0.21, 0.054, -1.455]\n",
    "biomass_value = 39.68\n",
    "biomass_df = pd.DataFrame()\n",
    "biomass_df['Metabolite'] = biomass_mets\n",
    "biomass_df['Value'] = bm_vals"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Using just the metabolites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_group_name</th>\n",
       "      <th>Metabolite</th>\n",
       "      <th>Formula</th>\n",
       "      <th>Intensity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Linearity_P1Ecoli_100xDil_1</td>\n",
       "      <td>12dgr120_c</td>\n",
       "      <td>C27H52O5</td>\n",
       "      <td>0.000073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Linearity_P1Ecoli_100xDil_1</td>\n",
       "      <td>12dgr140_c</td>\n",
       "      <td>C31H60O5</td>\n",
       "      <td>0.000010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Linearity_P1Ecoli_100xDil_1</td>\n",
       "      <td>12dgr141_c</td>\n",
       "      <td>C31H56O5</td>\n",
       "      <td>0.000010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Linearity_P1Ecoli_100xDil_1</td>\n",
       "      <td>12dgr160_c</td>\n",
       "      <td>C35H68O5</td>\n",
       "      <td>0.000227</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Linearity_P1Ecoli_100xDil_1</td>\n",
       "      <td>12dgr161_c</td>\n",
       "      <td>C35H64O5</td>\n",
       "      <td>0.000164</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10388</th>\n",
       "      <td>Linearity_P1Ecoli_1xDil_3</td>\n",
       "      <td>xdp_c</td>\n",
       "      <td>C10H14N4O12P2</td>\n",
       "      <td>0.000289</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10389</th>\n",
       "      <td>Linearity_P1Ecoli_1xDil_3</td>\n",
       "      <td>xmp_c</td>\n",
       "      <td>C10H13N4O9P1</td>\n",
       "      <td>0.000018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10390</th>\n",
       "      <td>Linearity_P1Ecoli_1xDil_3</td>\n",
       "      <td>xtp_c</td>\n",
       "      <td>C10H15N4O15P3</td>\n",
       "      <td>0.000065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10391</th>\n",
       "      <td>Linearity_P1Ecoli_1xDil_3</td>\n",
       "      <td>xtsn_c</td>\n",
       "      <td>C10H12N4O6</td>\n",
       "      <td>0.000176</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10392</th>\n",
       "      <td>Linearity_P1Ecoli_1xDil_3</td>\n",
       "      <td>zn2_c</td>\n",
       "      <td>Zn</td>\n",
       "      <td>0.000010</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10393 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 sample_group_name  Metabolite        Formula  Intensity\n",
       "0      Linearity_P1Ecoli_100xDil_1  12dgr120_c       C27H52O5   0.000073\n",
       "1      Linearity_P1Ecoli_100xDil_1  12dgr140_c       C31H60O5   0.000010\n",
       "2      Linearity_P1Ecoli_100xDil_1  12dgr141_c       C31H56O5   0.000010\n",
       "3      Linearity_P1Ecoli_100xDil_1  12dgr160_c       C35H68O5   0.000227\n",
       "4      Linearity_P1Ecoli_100xDil_1  12dgr161_c       C35H64O5   0.000164\n",
       "...                            ...         ...            ...        ...\n",
       "10388    Linearity_P1Ecoli_1xDil_3       xdp_c  C10H14N4O12P2   0.000289\n",
       "10389    Linearity_P1Ecoli_1xDil_3       xmp_c   C10H13N4O9P1   0.000018\n",
       "10390    Linearity_P1Ecoli_1xDil_3       xtp_c  C10H15N4O15P3   0.000065\n",
       "10391    Linearity_P1Ecoli_1xDil_3      xtsn_c     C10H12N4O6   0.000176\n",
       "10392    Linearity_P1Ecoli_1xDil_3       zn2_c             Zn   0.000010\n",
       "\n",
       "[10393 rows x 4 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "normalization.lim_tsi_norm(biomass_df['Metabolite'], intensities_Ecoli, columnname='Intensity')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Also using the stoichiometric coefficients"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_group_name</th>\n",
       "      <th>Metabolite</th>\n",
       "      <th>Formula</th>\n",
       "      <th>Intensity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Linearity_P1Ecoli_100xDil_1</td>\n",
       "      <td>12dgr120_c</td>\n",
       "      <td>C27H52O5</td>\n",
       "      <td>0.000896</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Linearity_P1Ecoli_100xDil_1</td>\n",
       "      <td>12dgr140_c</td>\n",
       "      <td>C31H60O5</td>\n",
       "      <td>0.000119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Linearity_P1Ecoli_100xDil_1</td>\n",
       "      <td>12dgr141_c</td>\n",
       "      <td>C31H56O5</td>\n",
       "      <td>0.000120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Linearity_P1Ecoli_100xDil_1</td>\n",
       "      <td>12dgr160_c</td>\n",
       "      <td>C35H68O5</td>\n",
       "      <td>0.002774</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Linearity_P1Ecoli_100xDil_1</td>\n",
       "      <td>12dgr161_c</td>\n",
       "      <td>C35H64O5</td>\n",
       "      <td>0.002004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10388</th>\n",
       "      <td>Linearity_P1Ecoli_1xDil_3</td>\n",
       "      <td>xdp_c</td>\n",
       "      <td>C10H14N4O12P2</td>\n",
       "      <td>0.007044</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10389</th>\n",
       "      <td>Linearity_P1Ecoli_1xDil_3</td>\n",
       "      <td>xmp_c</td>\n",
       "      <td>C10H13N4O9P1</td>\n",
       "      <td>0.000447</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10390</th>\n",
       "      <td>Linearity_P1Ecoli_1xDil_3</td>\n",
       "      <td>xtp_c</td>\n",
       "      <td>C10H15N4O15P3</td>\n",
       "      <td>0.001586</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10391</th>\n",
       "      <td>Linearity_P1Ecoli_1xDil_3</td>\n",
       "      <td>xtsn_c</td>\n",
       "      <td>C10H12N4O6</td>\n",
       "      <td>0.004290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10392</th>\n",
       "      <td>Linearity_P1Ecoli_1xDil_3</td>\n",
       "      <td>zn2_c</td>\n",
       "      <td>Zn</td>\n",
       "      <td>0.000245</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10393 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 sample_group_name  Metabolite        Formula  Intensity\n",
       "0      Linearity_P1Ecoli_100xDil_1  12dgr120_c       C27H52O5   0.000896\n",
       "1      Linearity_P1Ecoli_100xDil_1  12dgr140_c       C31H60O5   0.000119\n",
       "2      Linearity_P1Ecoli_100xDil_1  12dgr141_c       C31H56O5   0.000120\n",
       "3      Linearity_P1Ecoli_100xDil_1  12dgr160_c       C35H68O5   0.002774\n",
       "4      Linearity_P1Ecoli_100xDil_1  12dgr161_c       C35H64O5   0.002004\n",
       "...                            ...         ...            ...        ...\n",
       "10388    Linearity_P1Ecoli_1xDil_3       xdp_c  C10H14N4O12P2   0.007044\n",
       "10389    Linearity_P1Ecoli_1xDil_3       xmp_c   C10H13N4O9P1   0.000447\n",
       "10390    Linearity_P1Ecoli_1xDil_3       xtp_c  C10H15N4O15P3   0.001586\n",
       "10391    Linearity_P1Ecoli_1xDil_3      xtsn_c     C10H12N4O6   0.004290\n",
       "10392    Linearity_P1Ecoli_1xDil_3       zn2_c             Zn   0.000245\n",
       "\n",
       "[10393 rows x 4 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "normalization.lim_tsi_norm(biomass_df, intensities_Ecoli, biomass_value=biomass_value, columnname='Intensity')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Amino acids\n",
    "same but using AAs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# maybe add L-Selenocysteine and L-Pyrrolysine\n",
    "amino_acids = ['ala__L_c', 'arg__L_c', 'asn__L_c', 'asp__L_c',\n",
    "               'cys__L_c', 'glu__L_c', 'gln__L_c', 'gly_c',\n",
    "               'his__L_c', 'ile__L_c', 'leu__L_c', 'lys__L_c',\n",
    "               'met__L_c', 'phe__L_c', 'pro__L_c', 'ser__L_c',\n",
    "               'thr_c', 'trp__L_c', 'tyr__L_c', 'val__L_c']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_group_name</th>\n",
       "      <th>Metabolite</th>\n",
       "      <th>Formula</th>\n",
       "      <th>Intensity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Linearity_P1Ecoli_100xDil_1</td>\n",
       "      <td>12dgr120_c</td>\n",
       "      <td>C27H52O5</td>\n",
       "      <td>0.000101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Linearity_P1Ecoli_100xDil_1</td>\n",
       "      <td>12dgr140_c</td>\n",
       "      <td>C31H60O5</td>\n",
       "      <td>0.000013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Linearity_P1Ecoli_100xDil_1</td>\n",
       "      <td>12dgr141_c</td>\n",
       "      <td>C31H56O5</td>\n",
       "      <td>0.000014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Linearity_P1Ecoli_100xDil_1</td>\n",
       "      <td>12dgr160_c</td>\n",
       "      <td>C35H68O5</td>\n",
       "      <td>0.000313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Linearity_P1Ecoli_100xDil_1</td>\n",
       "      <td>12dgr161_c</td>\n",
       "      <td>C35H64O5</td>\n",
       "      <td>0.000226</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10388</th>\n",
       "      <td>Linearity_P1Ecoli_1xDil_3</td>\n",
       "      <td>xdp_c</td>\n",
       "      <td>C10H14N4O12P2</td>\n",
       "      <td>0.000394</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10389</th>\n",
       "      <td>Linearity_P1Ecoli_1xDil_3</td>\n",
       "      <td>xmp_c</td>\n",
       "      <td>C10H13N4O9P1</td>\n",
       "      <td>0.000025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10390</th>\n",
       "      <td>Linearity_P1Ecoli_1xDil_3</td>\n",
       "      <td>xtp_c</td>\n",
       "      <td>C10H15N4O15P3</td>\n",
       "      <td>0.000089</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10391</th>\n",
       "      <td>Linearity_P1Ecoli_1xDil_3</td>\n",
       "      <td>xtsn_c</td>\n",
       "      <td>C10H12N4O6</td>\n",
       "      <td>0.000240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10392</th>\n",
       "      <td>Linearity_P1Ecoli_1xDil_3</td>\n",
       "      <td>zn2_c</td>\n",
       "      <td>Zn</td>\n",
       "      <td>0.000014</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10393 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 sample_group_name  Metabolite        Formula  Intensity\n",
       "0      Linearity_P1Ecoli_100xDil_1  12dgr120_c       C27H52O5   0.000101\n",
       "1      Linearity_P1Ecoli_100xDil_1  12dgr140_c       C31H60O5   0.000013\n",
       "2      Linearity_P1Ecoli_100xDil_1  12dgr141_c       C31H56O5   0.000014\n",
       "3      Linearity_P1Ecoli_100xDil_1  12dgr160_c       C35H68O5   0.000313\n",
       "4      Linearity_P1Ecoli_100xDil_1  12dgr161_c       C35H64O5   0.000226\n",
       "...                            ...         ...            ...        ...\n",
       "10388    Linearity_P1Ecoli_1xDil_3       xdp_c  C10H14N4O12P2   0.000394\n",
       "10389    Linearity_P1Ecoli_1xDil_3       xmp_c   C10H13N4O9P1   0.000025\n",
       "10390    Linearity_P1Ecoli_1xDil_3       xtp_c  C10H15N4O15P3   0.000089\n",
       "10391    Linearity_P1Ecoli_1xDil_3      xtsn_c     C10H12N4O6   0.000240\n",
       "10392    Linearity_P1Ecoli_1xDil_3       zn2_c             Zn   0.000014\n",
       "\n",
       "[10393 rows x 4 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "normalization.lim_tsi_norm(amino_acids, intensities_Ecoli, columnname='Intensity')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Probabilistic Quotient Normalization\n",
    "This method [(Dieterle et. al, 2006)](https://pubs.acs.org/doi/10.1021/ac051632c) adjusts for dilutions so we would use it to compare different dilutions of the same sample, e.g. E. coli 1x, 10x, 100x etc\n",
    "\n",
    "**0)** Put all the corresponding samples into one dataframe. Because of the architecture I'd say columns correspond to sample, rows to metabolites. Use suitable values for missing metabolites, probably NA is fine but check that this is conform with the mean/median functions\n",
    "\n",
    "That one is 0 because it's still pre-processing. Here is the logic of the method\n",
    "\n",
    "**1)** tsi correct each sample separately\n",
    "\n",
    "**2)** Set up a QC vector (or additional column) with the mean/median for each metabolite over all samples\n",
    "\n",
    "**3)** Divide each value for each sample with the correscponding QC value\n",
    "\n",
    "**4)** Get the mean/median over all metabolites for each sample -> This is your dilution factor for this sample\n",
    "\n",
    "**5)** Take the values from 1) and multiply each value with the dilution factor corresponding to this sample\n",
    "\n",
    "And to finish that, export it like the input\n",
    "\n",
    "**6)** Copy the input df and insert the new values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_group_name</th>\n",
       "      <th>Metabolite</th>\n",
       "      <th>Formula</th>\n",
       "      <th>Intensity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Linearity_P1Ecoli_100xDil_1</td>\n",
       "      <td>12dgr120_c</td>\n",
       "      <td>C27H52O5</td>\n",
       "      <td>1.401056e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Linearity_P1Ecoli_100xDil_1</td>\n",
       "      <td>12dgr140_c</td>\n",
       "      <td>C31H60O5</td>\n",
       "      <td>1.866019e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Linearity_P1Ecoli_100xDil_1</td>\n",
       "      <td>12dgr141_c</td>\n",
       "      <td>C31H56O5</td>\n",
       "      <td>1.874573e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Linearity_P1Ecoli_100xDil_1</td>\n",
       "      <td>12dgr160_c</td>\n",
       "      <td>C35H68O5</td>\n",
       "      <td>4.339788e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Linearity_P1Ecoli_100xDil_1</td>\n",
       "      <td>12dgr161_c</td>\n",
       "      <td>C35H64O5</td>\n",
       "      <td>3.135457e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10388</th>\n",
       "      <td>Linearity_P1Ecoli_1xDil_3</td>\n",
       "      <td>xdp_c</td>\n",
       "      <td>C10H14N4O12P2</td>\n",
       "      <td>5.773747e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10389</th>\n",
       "      <td>Linearity_P1Ecoli_1xDil_3</td>\n",
       "      <td>xmp_c</td>\n",
       "      <td>C10H13N4O9P1</td>\n",
       "      <td>3.660085e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10390</th>\n",
       "      <td>Linearity_P1Ecoli_1xDil_3</td>\n",
       "      <td>xtp_c</td>\n",
       "      <td>C10H15N4O15P3</td>\n",
       "      <td>1.300094e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10391</th>\n",
       "      <td>Linearity_P1Ecoli_1xDil_3</td>\n",
       "      <td>xtsn_c</td>\n",
       "      <td>C10H12N4O6</td>\n",
       "      <td>3.516230e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10392</th>\n",
       "      <td>Linearity_P1Ecoli_1xDil_3</td>\n",
       "      <td>zn2_c</td>\n",
       "      <td>Zn</td>\n",
       "      <td>2.005388e-06</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10393 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 sample_group_name  Metabolite        Formula     Intensity\n",
       "0      Linearity_P1Ecoli_100xDil_1  12dgr120_c       C27H52O5  1.401056e-06\n",
       "1      Linearity_P1Ecoli_100xDil_1  12dgr140_c       C31H60O5  1.866019e-07\n",
       "2      Linearity_P1Ecoli_100xDil_1  12dgr141_c       C31H56O5  1.874573e-07\n",
       "3      Linearity_P1Ecoli_100xDil_1  12dgr160_c       C35H68O5  4.339788e-06\n",
       "4      Linearity_P1Ecoli_100xDil_1  12dgr161_c       C35H64O5  3.135457e-06\n",
       "...                            ...         ...            ...           ...\n",
       "10388    Linearity_P1Ecoli_1xDil_3       xdp_c  C10H14N4O12P2  5.773747e-05\n",
       "10389    Linearity_P1Ecoli_1xDil_3       xmp_c   C10H13N4O9P1  3.660085e-06\n",
       "10390    Linearity_P1Ecoli_1xDil_3       xtp_c  C10H15N4O15P3  1.300094e-05\n",
       "10391    Linearity_P1Ecoli_1xDil_3      xtsn_c     C10H12N4O6  3.516230e-05\n",
       "10392    Linearity_P1Ecoli_1xDil_3       zn2_c             Zn  2.005388e-06\n",
       "\n",
       "[10393 rows x 4 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "normalization.pqn_norm(intensities_Ecoli, 'sample_group_name', 'Intensity', 'median')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_group_name</th>\n",
       "      <th>Metabolite</th>\n",
       "      <th>Formula</th>\n",
       "      <th>Intensity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Linearity_P1Ecoli_100xDil_1</td>\n",
       "      <td>12dgr120_c</td>\n",
       "      <td>C27H52O5</td>\n",
       "      <td>1.386942e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Linearity_P1Ecoli_100xDil_1</td>\n",
       "      <td>12dgr140_c</td>\n",
       "      <td>C31H60O5</td>\n",
       "      <td>1.847222e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Linearity_P1Ecoli_100xDil_1</td>\n",
       "      <td>12dgr141_c</td>\n",
       "      <td>C31H56O5</td>\n",
       "      <td>1.855689e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Linearity_P1Ecoli_100xDil_1</td>\n",
       "      <td>12dgr160_c</td>\n",
       "      <td>C35H68O5</td>\n",
       "      <td>4.296071e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Linearity_P1Ecoli_100xDil_1</td>\n",
       "      <td>12dgr161_c</td>\n",
       "      <td>C35H64O5</td>\n",
       "      <td>3.103872e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10388</th>\n",
       "      <td>Linearity_P1Ecoli_1xDil_3</td>\n",
       "      <td>xdp_c</td>\n",
       "      <td>C10H14N4O12P2</td>\n",
       "      <td>6.273600e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10389</th>\n",
       "      <td>Linearity_P1Ecoli_1xDil_3</td>\n",
       "      <td>xmp_c</td>\n",
       "      <td>C10H13N4O9P1</td>\n",
       "      <td>3.976950e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10390</th>\n",
       "      <td>Linearity_P1Ecoli_1xDil_3</td>\n",
       "      <td>xtp_c</td>\n",
       "      <td>C10H15N4O15P3</td>\n",
       "      <td>1.412648e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10391</th>\n",
       "      <td>Linearity_P1Ecoli_1xDil_3</td>\n",
       "      <td>xtsn_c</td>\n",
       "      <td>C10H12N4O6</td>\n",
       "      <td>3.820641e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10392</th>\n",
       "      <td>Linearity_P1Ecoli_1xDil_3</td>\n",
       "      <td>zn2_c</td>\n",
       "      <td>Zn</td>\n",
       "      <td>2.179001e-06</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10393 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 sample_group_name  Metabolite        Formula     Intensity\n",
       "0      Linearity_P1Ecoli_100xDil_1  12dgr120_c       C27H52O5  1.386942e-06\n",
       "1      Linearity_P1Ecoli_100xDil_1  12dgr140_c       C31H60O5  1.847222e-07\n",
       "2      Linearity_P1Ecoli_100xDil_1  12dgr141_c       C31H56O5  1.855689e-07\n",
       "3      Linearity_P1Ecoli_100xDil_1  12dgr160_c       C35H68O5  4.296071e-06\n",
       "4      Linearity_P1Ecoli_100xDil_1  12dgr161_c       C35H64O5  3.103872e-06\n",
       "...                            ...         ...            ...           ...\n",
       "10388    Linearity_P1Ecoli_1xDil_3       xdp_c  C10H14N4O12P2  6.273600e-05\n",
       "10389    Linearity_P1Ecoli_1xDil_3       xmp_c   C10H13N4O9P1  3.976950e-06\n",
       "10390    Linearity_P1Ecoli_1xDil_3       xtp_c  C10H15N4O15P3  1.412648e-05\n",
       "10391    Linearity_P1Ecoli_1xDil_3      xtsn_c     C10H12N4O6  3.820641e-05\n",
       "10392    Linearity_P1Ecoli_1xDil_3       zn2_c             Zn  2.179001e-06\n",
       "\n",
       "[10393 rows x 4 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "normalization.pqn_norm(intensities_Ecoli, 'sample_group_name', 'Intensity', 'mean')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bfair",
   "language": "python",
   "name": "bfair"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
